File: scripts/generate-placeholders/ramps.py
"""ramps.py — load the colour ramps from the single source of truth, js/ramps.js.

The overlay builders (generate_overlays.py, build_real_ndvi.py, build_real_thermal.py) call
`load_ramps()` instead of defining their own stops, so the Python LUTs and the JavaScript
legends/inspect reverse-lookup all come from the same definition. js/ramps.js is formatted
one `[position, "#hex"]` pair per line specifically to keep this parser robust.
"""

import os
import re

_PAIR = re.compile(r'\[\s*([\d.]+)\s*,\s*"(#[0-9a-fA-F]{6})"\s*\]')
_HEADER = re.compile(r"(\w+)\s*:\s*\[\s*$")


def load_ramps(path=None):
    """Return {'viridis': [(pos, '#hex'), ...], 'inferno': [...]} from js/ramps.js."""
    if path is None:
        path = os.path.join(os.path.dirname(__file__), "..", "..", "js", "ramps.js")
    ramps, current = {}, None
    with open(path) as fh:
        for line in fh:
            header = _HEADER.search(line)
            if header and '"' not in line:  # a ramp header line, e.g. "  viridis: ["
                current = header.group(1)
                ramps[current] = []
                continue
            pair = _PAIR.search(line)
            if pair and current is not None:
                ramps[current].append((float(pair.group(1)), pair.group(2)))
    if "viridis" not in ramps or "inferno" not in ramps:
        raise ValueError(f"Could not parse viridis/inferno ramps from {path}")
    return ramps


if __name__ == "__main__":
    for name, stops in load_ramps().items():
        print(name, stops)

File: js/ramps.js
// ramps.js — SINGLE SOURCE OF TRUTH for the colour ramps.
//
// Consumed by js/config.js (legend gradients), js/sample.js (reverse-lookup for
// click-to-inspect), and the Python overlay builders via
// scripts/generate-placeholders/ramps.py, which parses THIS file. There is no longer a
// duplicated ramp definition to keep in sync.
//
// Format is intentionally simple — one [position, "#hex"] pair per line — so the Python
// regex parser stays robust. Positions are 0..1 and match the LUTs used to colourise the
// exported PNGs, which is what makes the inspect reverse-lookup accurate.

export const RAMPS = {
  viridis: [
    [0.0, '#440154'],
    [0.25, '#3b528b'],
    [0.5, '#21918c'],
    [0.75, '#5ec962'],
    [1.0, '#fde725'],
  ],
  inferno: [
    [0.0, '#000004'],
    [0.25, '#420a68'],
    [0.5, '#932667'],
    [0.75, '#dd513a'],
    [0.9, '#fca50a'],
    [1.0, '#fcffa4'],
  ],
};

// Presentation helper (browser only): build a CSS `linear-gradient(...)` from a ramp's
// [position, hex] stops, so the legend gradients and the story-card swatch derive from the
// same source as the raster LUTs. Written to not match the Python parser's pair/header
// regexes above, so it never interferes with load_ramps().
export function rampGradientCss(stops, angle = '90deg') {
  const parts = stops.map((s) => `${s[1]} ${(s[0] * 100).toFixed(1)}%`).join(', ');
  return `linear-gradient(${angle}, ${parts})`;
}
